systems/ps3.ixx

```diff
@@ -0,0 +1,262 @@
+import utils.misc;
+import utils.strings;
+
+
+
+namespace gpsxre
+{
+
+export class SystemPS3 : public System
+{
+public:
+	std::string getName() override
+	{
+		return "PS3";
+	}
+
+
+	Type getType() override
+	{
+		return Type::ISO;
+	}
+
+
+	void printInfo(std::ostream &os, SectorReader *sector_reader, const std::filesystem::path &) const override
+	{
+		iso9660::PrimaryVolumeDescriptor pvd;
+		if(!iso9660::Browser::findDescriptor((iso9660::VolumeDescriptor &)pvd, sector_reader, iso9660::VolumeDescriptorType::PRIMARY))
+			return;
+		auto root_directory = iso9660::Browser::rootDirectory(sector_reader, pvd);
+
+		auto ps3_disc_sfb = loadSFB(root_directory, "PS3_DISC.SFB");
+		auto param_sfo = loadSFO(root_directory, "PS3_GAME/PARAM.SFO");
+
+		auto it = ps3_disc_sfb.find("VERSION");
+		if(it != ps3_disc_sfb.end())
+			os << std::format("  version: {}", it->second) << std::endl;
+		else
+		{
+			it = param_sfo.find("VERSION");
+			if(it != param_sfo.end())
+				os << std::format("  version: {}", it->second) << std::endl;
+		}
+
+		std::string title_id;
+		it = ps3_disc_sfb.find("TITLE_ID");
+		if(it != ps3_disc_sfb.end())
+			title_id = it->second;
+		else
+		{
+			it = param_sfo.find("TITLE_ID");
+			if(it != param_sfo.end())
+				title_id = it->second;
+		}
+
+		auto serial = deduceSerial(title_id);
+		if(!serial.first.empty() && !serial.second.empty())
+			os << std::format("  serial: {}-{}", serial.first, serial.second) << std::endl;
+
+		auto fw = deduceFirmware(root_directory, "PS3_UPDATE/PS3UPDAT.PUP");
+		if(!fw.empty())
+			os << std::format("  firmware: {}", fw) << std::endl;
+	}
+
+private:
+	static constexpr std::array<uint8_t, 4> _SFB_MAGIC = { 0x2E, 0x53, 0x46, 0x42 };
+	static constexpr std::array<uint8_t, 4> _SFO_MAGIC = { 0x00, 0x50, 0x53, 0x46 };
+
+	std::map<std::string, std::string> loadSFB(std::shared_ptr<iso9660::Entry> root_directory, std::string sfb_file) const
+	{
+		std::map<std::string, std::string> sfb;
+
+		auto sfb_entry = root_directory->subEntry(sfb_file);
+		if(sfb_entry)
+		{
+			auto data = sfb_entry->read();
+			if(data.size() < 0x200)
+				return sfb;
+
+			if(!std::equal(data.begin(), data.begin() + 4, _SFB_MAGIC.begin()))
+				return sfb;
+
+			size_t pos = 0x20;
+			std::string key(data.begin() + pos, data.begin() + pos + 0x10);
+			std::string value;
+			uint32_t offset;
+			uint32_t length;
+			while(pos < 0x1E0)
+			{
+				offset = (static_cast<uint32_t>(data[pos + 0x10]) << 24) |
+						 (static_cast<uint32_t>(data[pos + 0x11]) << 16) |
+						 (static_cast<uint32_t>(data[pos + 0x12]) << 8) |
+						  static_cast<uint32_t>(data[pos + 0x13]);
+				length = (static_cast<uint32_t>(data[pos + 0x14]) << 24) |
+					     (static_cast<uint32_t>(data[pos + 0x15]) << 16) |
+					     (static_cast<uint32_t>(data[pos + 0x16]) << 8) |
+					      static_cast<uint32_t>(data[pos + 0x17]);
+
+				if(data.size() < offset + length)
+					return sfb;
+
+				value.assign(reinterpret_cast<char*>(&data[offset]), length);
```

Comment:
I prefer having no casts and I think most of these will be resolved if properly implemented using structures.
But if cast is still required, I like C-style because it's more tidy in my opinion.